packages/server/routes/files.js

```diff
@@ -43,20 +51,79 @@ const { extract } = require("@saltcorn/admin-models/models/backup");
 const router = new Router();
 module.exports = router;
 
-/**
- * Edit file Role form
- * @param {*} file
- * @param {*} roles
- * @param {*} req
- * @returns {Form}
- */
-const editFileRoleForm = (file, roles, req) =>
-  editRoleForm({
-    url: `/files/setrole/${file.path_to_serve}`,
-    current_role: file.min_role_read,
-    roles,
-    req,
+const send_files_picker = async (folder, noSubdirs, inputId, req, res) => {
+  res.set("SaltcornModalWidth", "1200px");
+  res.sendWrap(req.__("Please select a file"), {
+    above: [
+      script({
+        src: `/static_assets/${db.connectObj.version_tag}/bundle.js`,
+        defer: true,
+      }),
+      script(
+        domReady(
+          `$("head").append('${link({
+            rel: "stylesheet",
+            href: `/static_assets/${db.connectObj.version_tag}/bundle.css`,
+          })}')`
+        )
+      ),
+      div({
+        id: "saltcorn-file-manager",
+        full_manager: "false",
+        folder: folder,
+        input_id: inputId,
+        ...(noSubdirs ? { no_subdirs: "true" } : {}),
+      }),
+    ],
   });
+};
+
+router.get(
+  "/picker",
+  error_catcher(async (req, res) => {
+    const { folder, input_id, no_subdirs } = req.query;
+    send_files_picker(folder, no_subdirs, input_id, req, res);
+  })
+);
+
+router.get(
+  "/visible_entries",
+  error_catcher(async (req, res) => {
+    const role = req.user?.role_id ? req.user.role_id : 100;
+    const { dir, no_subdirs } = req.query;
+    const noSubdirs = no_subdirs === "true";
+    const safeDir = File.normalise(dir || "/");
+    const rows = (
+      await File.find({ folder: dir }, { orderBy: "filename" })
+    ).filter((f) => {
+      if (noSubdirs && f.isDirectory) return false;
+      else return role <= f.min_role_read;
```

Comment:
or user is file owner?

Reply:
I'm checking the file user_id now. I believe we can't use any row formulas because we're on the file level